src/preamble.js

```diff
@@ -317,10 +317,25 @@ function allocate(slab, types, allocator, ptr) {
   var singleType = typeof types === 'string' ? types : null;
 
   var ret;
-  if (allocator == ALLOC_NONE) {
-    ret = ptr;
-  } else {
-    ret = [typeof _malloc === 'function' ? _malloc : staticAlloc, stackAlloc, staticAlloc, dynamicAlloc][allocator === undefined ? ALLOC_STATIC : allocator](Math.max(size, singleType ? 1 : types.length));
+  var sz = Math.max(size, singleType ? 1 : types.length);
+  switch(allocator) {
+    case ALLOC_NONE:
+      ret = ptr;
+      break;
+    case ALLOC_DYNAMIC:
+      ret = dynamicAlloc(sz);
+      break;
+    case ALLOC_STACK:
+      ret = stackAlloc(sz);
+      break;
+    case ALLOC_NORMAL:
+      if (typeof _malloc === 'function') {
+        ret = _malloc(sz);
+        break;
+      }
+      // Fall through to perform a static allocation if malloc is not yet available
+    default: // case ALLOC_STATIC:
+      ret = staticAlloc(sz);
   }
```

Comment:
this version is larger, and I don't see the benefit?

Reply:
The reason for this change is that earlier code referenced `stackAlloc`, which is exported from asm.js/wasm module, and after no longer statically declaring it, the `allocate()` function would fail at undefined reference to `stackAlloc`. This code is larger by ~150 bytes, but this also avoid temp garbage creation, so that is worth it to my view.

I could have added `typeof stackAlloc === 'function'` checks, but that makes the code really complex, and the temp garbage variant there really bugged me.

Ideally in the future we'll work towards removing `allocate()` altogether, and call sites should know what/where they are allocating, so preferably we will do away with this function in the future anyways. I don't think any user code should ever be calling `allocate()`, so this should be doable.

Reply:
I see, thanks.

Then this change should no longer be necessary after the fixes for closure/metadce, beecause then we will know stackAlloc exists in those optimizers?

Reply:
This is still needed - before the code was

```javascript
function allocate(...) {
  ...
  ret = [..., stackAlloc, ...](...);
  ...
}
...
var stackAlloc = Module['stackAlloc'] = asm['stackAlloc'];
```

which parsed, but `stackAlloc` would be `undefined` in all calls to `allocate()` until execution reached to receiving the exports. (If anything attempted to do a `stackAlloc()`, it'd throw an exception trying to call `undefined()`, but that does not happen).

After this change we no longer have the line `var stackAlloc = Module['stackAlloc'] = asm['stackAlloc'];`, at which point referencing `stackAlloc` is not allowed, but throws immediately. I could do `typeof stackAlloc == 'function' ? stackAlloc : 0` there, but thought it is just better to simplify code.

Reply:
If we have to, this version with `ifdef`ing seems ok, but could we even more simply do in an earlier toplevel location
```
#if !DECLARE_..
var stackAlloc; // declare it explicitly for allocate()
#endif 
```
That would give it a value (undefined) so allocate does not crash before the asm module returns the implementation?